Add AfterDark

goakley opened the pull request:
Hello!

I noticed that AfterDark was not in the list of supported websites.
I have implemented a basic AfterDark `WebsiteService` that successfully authenticates and uploads images.
Since I could not find any contributor guidelines, I tried my best to copy your approach to implementing the different components, but I may have missed some best practices.

Thank you in advance for considering this code.

mvdicarlo commented:
I will try to take a look at this implementation within the week.
And yeah, I have been bad about setting up a guideline around the app since I kind of gave of that people would contribute after going it solo for nearly 3 years.

Hopefully it wasn't too much of a pain in the ass.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+    tags = formData.defaults.tags.tags;
+  }
+  if (tags.length < 2) {
+    return [['AfterDark requires at least two tags', { website: 'AfterDark', value: tags.length }]];
```
Incorrect format.
Check weasyl.service.ts to see how tag string should be (because of translation).

goakley replied:
Switched to correct string

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+  }
+  const collections = ((formData['AfterDark'] || {}).options || {}).folders || [];
+  if (collections.length < 1) {
+    return [['AfterDark requires at least one collection folder', { website: 'AfterDark', value: collections.length }]];
```
Make single string. Setting a website + value object requires some kind of translation equivalent in translation files.

goakley replied:
I understand the translation structure now - witched to a single string.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+  if (collections.length < 1) {
+    return [['AfterDark requires at least one collection folder', { website: 'AfterDark', value: collections.length }]];
+  }
+  return [];
```
Preferred to have a array set at the top and then .push into it as issues are found.
Not just return [[...]]

goakley replied:
Now I get it - done.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+    if (response.statusCode !== 200) {
+      throw new Error(`Bad response getting csrf token ${response.statusCode}`);
+    }
+    const match = response.body.match(new RegExp(`name="csrfmiddlewaretoken" value="([^"]+)"`));
```
Declare regex inside of /<regex>/

goakley replied:
Cool!

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+      }`
+    });
+    var cookies = await getCookies(profileId, this.BASE_URL);
+    const response = await got.gotPostJSON(`${this.BASE_URL}/graphql/`, body, this.BASE_URL, cookies, profileId, {
```
I am like 99% certain a new get.gotPostJSON did not need to be made.
Both Request and Got should be fine with sending non-formdata.
`await got.post(`${this.BASE_URL}/graphql/`, null, this.BASE_URL, cookies, profileId, {
    body,
    json: true,
    headers: {...}
})`

https://github.com/request/request#requestoptions-callback

goakley replied:
Ahh, I was thrown by having to set the second param to `{}`.  Setting it to null is an error because gotPost doesn't check if the body is null.  Removed getPostJSON.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+    if (!(responseBody.data || {}).userViewer) {
+      return null;
+    }
+    const collections = ((((responseBody.data || {}).userViewer || {}).user || {}).collectionSet || {}).edges || [];
+    return {
+      'userId': (((responseBody.data || {}).userViewer || {}).user || {}).id || '',
+      'username': (((responseBody.data || {}).userViewer || {}).user || {}).username || '',
+      'collections': collections.map(c => c.node),
```
All of this chaining is ugly and error prone.
Look at website-validatgor.helper.ts (getTags) as an example and use dotProp.get(...)

goakley replied:
This library is great!  Done.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  public async post(submission: Submission, postData: SubmissionPostData): Promise<PostResult> {
+    if (submission.submissionType !== SubmissionType.SUBMISSION) {
+      throw new Error('Submissions only.')
+    }
```
Redundant.
No post type of journal could get here as you have not specified a journal type form.

goakley replied:
check

goakley replied:
Done.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+    if (submission.submissionType !== SubmissionType.SUBMISSION) {
+      throw new Error('Submissions only.')
+    }
+    if (postData.typeOfSubmission != TypeOfSubmission.ART) {
+      throw new Error('Art submissions only.');
+    }
```
Remove.
Do this kind of check in the validation function.

Check how twitter.service.ts checks for file type on line 21 inside of submissionValidate.

goakley replied:
check

goakley replied:
Replaced this check with a check in submissionValidate.

mvdicarlo commented on src/app/websites/website-services/afterdark/components/afterdark-submission-form/afterdark-submission-form.component.html:
```diff
@@ -0,0 +1,13 @@
+<form [formGroup]="formGroup">
+  <div formGroupName="options" class="row no-gutters section">
+    <div class="row no-gutters">
+      <div class="col-md col-lg col-xl">
+        <div class="row no-gutters">
+          <div class="col-md col-lg col-xl">
+            <folder-select formControlName="folders" [folders]="folders" [multiple]="true"></folder-select>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+</form>
```
Where is the Tags input?
See inkbunny-submission-form.component.html and inkbunny-submission-form.component.ts

goakley replied:
Added a tags input to this component.

mvdicarlo commented on src/app/websites/website-services/afterdark/components/afterdark-submission-form/afterdark-submission-form.component.ts:
```diff
@@ -0,0 +1,37 @@
+import { Component, OnInit, Injector, forwardRef, AfterViewInit } from '@angular/core';
+import { BaseWebsiteSubmissionForm, HOST_DATA } from 'src/app/websites/components/base-website-submission-form/base-website-submission-form.component';
+import { TagConfig } from 'src/app/utils/components/tag-input/tag-input.component';
+import { Folder } from 'src/app/websites/interfaces/folder.interface';
+
+@Component({
+  selector: 'afterdark-submission-form',
+  templateUrl: './afterdark-submission-form.component.html',
+  styleUrls: ['./afterdark-submission-form.component.css'],
+  providers: [{ provide: BaseWebsiteSubmissionForm, useExisting: forwardRef(() => AfterDarkSubmissionForm) }],
+  host: HOST_DATA
+})
+export class AfterDarkSubmissionForm extends BaseWebsiteSubmissionForm implements OnInit, AfterViewInit {
```
Tags

goakley replied:
Added a tags input to this component.

mvdicarlo commented on src/typings.d.ts:
```diff
@@ -14,6 +14,7 @@ declare var got: {
   patch(url: string, formData: any, cookieUrl: string, cookies: any[], options?: any): Promise<{ error?: any, success?: { body: any, response: any } }>;
   post(url: string, formData: any, cookieUrl: string, cookies: any[], options?: any): Promise<{ error?: any, success?: { body: any, response: any } }>;
   gotPost(url: string, formData: any, cookieUrl: string, cookies: any[], options?: any): Promise<any>;
+  gotPostJSON(url: string, body: any, cookieUrl: string, cookies: any[], profileId: string, options?: any): Promise<any>;
```
Potential removal if found to be unnecessary.

goakley replied:
Removed.

goakley commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+  if (tags.length < 2) {
+    return [['AfterDark requires at least two tags', { website: 'AfterDark', value: tags.length }]];
+  }
+  const collections = ((formData['AfterDark'] || {}).options || {}).folders || [];
```
Done.

goakley commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,201 @@
+      "nsfw": postData.rating !== SubmissionRating.GENERAL && postData.rating !== SubmissionRating.MATURE,
+      "claimedByUser": true,
+    });
+    var formData = {
```
Switched all `var`s to `let`s.

goakley commented on src/electron-src/src/got-request.js:
```diff
@@ -216,6 +216,49 @@ exports.gotPost = function gotPost(url, formData, cookieUrl, cookies, options) {
+        resolve(err); // got seems to throw a lot despite a successful post
+      });
+  });
+};
```
Not necessary!  \o/

goakley commented:
Thank you for the thorough review, and sorry for missing so many obvious little things!  I believe I have addressed all the issues.  The code is still working as expected in manual testing.

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,199 @@
+function submissionValidate(submission: Submission, formData: SubmissionFormData): any[] {
+  const problems: any[] = [];
+  if (!supportsFileType(submission.fileInfo, ACCEPTED_FILES)) {
+    problems.push(['Does not support file format', { website: 'Twitter', value: submission.fileInfo.type }]);
```
```suggestion
    problems.push(['Does not support file format', { website: 'AfterDark', value: submission.fileInfo.type }]);
```

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,199 @@
+    if (!match) {
+      throw new Error('Could not find csrf token in body');
+    }
+    const csrfToken = match[1];
+    return csrfToken;
```
```suggestion
    return match[1];
```

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,199 @@
+      }`
+    });
+    let cookies = await getCookies(profileId, this.BASE_URL);
+    const response = await got.gotPost(`${this.BASE_URL}/graphql/`, {}, this.BASE_URL, cookies, {
+      body,
+      headers: {
+        'content-type': 'application/json',
+        'X-CSRFToken': csrfToken,
+      },
+    });
```
Good to see you were able to get this to work.
Just want to verify that you tested it and it was working as expected.

goakley replied:
Yep!  This call is proof that the user is still logged in.

![login](https://user-images.githubusercontent.com/1375517/92552226-00361200-f215-11ea-80ee-abb8a2c14709.png)

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,199 @@
+    };
+    const csrfToken = await this.getCsrfToken(profileId);
+    const userData = await this.getUserData(profileId, csrfToken);
+    if (!userData) {
+      return returnValue;
+    }
+    this._profileManager.storeData(profileId, AfterDark.name, userData);
+    returnValue.username = userData.username;
+    returnValue.status = LoginStatus.LOGGED_IN;
+    return returnValue;
```
```suggestion
    if (userData) {
          this._profileManager.storeData(profileId, AfterDark.name, userData);
          returnValue.username = userData.username;
          returnValue.status = LoginStatus.LOGGED_IN;
    }
    return returnValue;
```

mvdicarlo commented on src/app/websites/website-services/afterdark/afterdark.service.ts:
```diff
@@ -0,0 +1,199 @@
+      variables,
+    };
+    formData[postData.primary.fileInfo.name] = fileFormData;
+    const response = await got.post(`${this.BASE_URL}/graphql/`, formData, this.BASE_URL, cookies, {
+      headers: {
+        'content-type': 'multipart/form-data',
+        'X-CSRFToken': csrfToken,
+      }
+    });
```
I thought an earlier commit said that graphql didn't accept multipart and only took json?
If it works its works, I just wanted to re-verify.

goakley replied:
The site is a bit weird - for posting an upload, it uses some hack to allow the large image body that involves taking form data, of which the usual graphql fields are form data elements.

An example taken directly from `afterdark.art/upload`: https://pastebin.com/raw/uxTGCGZQ

And result:
![2020-09-08-205816_833x805_scrot](https://user-images.githubusercontent.com/1375517/92552666-0e386280-f216-11ea-89e9-22fdbd684ead.png)

mvdicarlo reviewed (commented):
Everything looks to be mostly in order.

I just wanted to verify some things I asked in the comments and for some minor code tweaks.
Once that is done I will merge it in and manually test it myself as well.

Hopefully my previous comments weren't too annoying to work with.

mvdicarlo commented:
Thanks for all the hard work in implementing this!
